### ISSUE TITLE
Improve bevy plugin

### ISSUE THREAD

**Issue by ErnWong** [ErnWong/crystalorb#14]
Right now, if we were to use the provided bevy plugin, we would need to write most of the game logic outside of bevy's ECS. It would be good to integrate crystalorb with bevy's ECS.

See other similar plugins for examples
- https://github.com/gschup/bevy_ggrs
- https://github.com/HouraiTeahouse/backroll-rs/tree/main/bevy_backroll

Also investigate any new or upcoming bevy ECS features that would make the plugin more ergonomic.

**Comment by Shatur** [comment]
Yes, It would be great to decouple world somehow. Because right now I should put all my networking logic into a single data structure.

**Comment by Shatur** [comment]
@ErnWong, do you have any design comes to mind? I would like to help.

**Comment by ErnWong** [comment]
Thanks for your interest! I've got some ideas that might be cool for us to try out. I'll see if I can do bit of investigation this weekend to flesh them out and will get back to you. (Feel free to ping me if I don't get back).

Ideally, it'll be nice to see if we can leave the core CrystalOrb crate more or less the same, and have the changes mostly reside on our bevy plugin crate side of things. CrystalOrb is designed around simulating two instances of the World at any given moment, which, if I understand correctly, is quite different to how backroll and ggrs does it, so it may be cleaner to structure our bevy integration approach differently to theirs.

### The Bevy Sandwich Idea

I might as well share what I have in mind, but feel free to suggest other ideas too!

If we went with CrystalOrb's current style of doing things, then we might try and implement the `crystalorb::world::World` trait for an entire `bevy_app::App` (or `bevy_ecs::world::World`) and we'll end up with a delicious bevy sandwich (maybe we should call the plugin `bevy_sandwich` :joy:):

- We have an outer `bevy_app::App`, consisting of
   - e.g. the users' game rendering systems
   - other game logic systems that doesn't need to be networked
   - entities and components to be displayed to the screen
   - the `bevy_crystalorb` plugin, which supplies:
       - a `crystalorb::client::Client<bevy_crystalorb::World>` bevy resource, which:
           - internally spawns two instances of `bevy_crystalorb::World`, each of which consists of:
              - an inner `bevy_app::App`, which consists of
                 - entities and components to be networked
                 - update systems and setup systems for running your networked game simulation logic
       - a `bevy_crystalorb::update_system`, which invokes `crystalorb::client::Client::update` and syncs `crystalorb::client::stage::Ready::display_state` with the outer bevy app's components.

The game developer would probably set up their game code the usual way, and tag/register their components and systems that they want CrystalOrb to network. Then, the CrystalOrb bevy plugin would hook the inner bevy apps to the correct components and systems, and act as a bridge between the inner and outer bevy apps.
       
Hmm, that might be a bit confusing. Maybe I should try and draw it out:

![image](https://user-images.githubusercontent.com/2609018/136183284-56bce36c-1722-46e6-a3e3-19ee051a567b.png)

Yikes, that looks even more confusing. This approach might sound a bit heavy-weight on paper, but so far, CrystalOrb's design decisions has been based around choosing the "cleaner" option over a more performant option. ("Clean" is very subjective though 😛 )

### Bevy Subworlds

Having bevy apps inside a bevy app sounds a bit like the bevy subworlds RFC (https://github.com/bevyengine/rfcs/pull/16). I haven't explored that RFC in too much detail, but I'm not sure if it'll be easy to make our ownership hierarchy clean using that approach (since I'm guessing we'll need some sort of circular reference between bevy and our crystalorb client?).

**Comment by Shatur** [comment]
Sounds interesting :) But why we need to spawn two worlds?

You may also find it interesting that the Naia network engine is also trying to implement integration for Bevy too: https://github.com/naia-rs/naia/pull/22
I thought you might be interested in looking at this implementation. I talked a bit with the author on Discord, he plans to complete the integration in a few weeks. Naia advantage is that it doesn't need a nightly compiler.

**Comment by Shatur** [comment]
> We simulate two worlds because whenever we want to perform a rollback:

Got it, sounds reasonable!

**Comment by jamescarterbell** [comment]
I actually think I figured out number 2, I was just being dumb, but we'll see soon.

> Just double checking - did you mean systems rather than plugins? (since if the game developer has already organised their game into plugins, they could pass in a plugin or group of plugins into our crystalorb plugin to initialise the inner world without having to wrap it inside some special builder function, and we'll be sweet!).

I do mean plugins! Since the inner world and outer world can essentially be thought of as different apps in this model, you really do need two plugins for any plugin that will interact with both the inner and outer worlds, since there's no way to directly access one from the other (they interact via displaystates, and *eventually* the outer world holds both inner worlds, but before then they're seperate).

I think the builder idea will work, but the only annoying thing is it will mean having to create your app kind of like this:

`
let app_factory = ||{
     App::build()
          .add_plugin(BevyCrystalOrbPhysicsSimulationPlugin)
         .app
}

App::build()
     .add_plugin(CrystalOrbClientPlugin::with_factory(app_factory))
`

Then the client plugin can run your builder twice, and take the world and scheduler from that app.

**Comment by ErnWong** [comment]
> I do mean plugins! 

Ah, cool!

> but the only annoying thing is it will mean having to create your app kind of like this:

I'm assuming we're wrapping it in a `app_factory` closure because we don't want to call the simulation plugin's `.build()` twice on the same plugin object? (I'm happy either way - just double checking the reason).

Btw, does this mean we need to refactor `crystalorb::client::Client::new()` to take in a world factory? (This is because we're currently initialising the world using `crystalorb::world::World::default()`, which I'm guessing won't play well with our `app_factory?`). Or, do you have other plans for injecting the plugin into the already created `crystalorb::world::World`?

Oh, I've got another idea! Rather than have an `app_factory` closure and refactoring `crystalorb::client::Client::new()`, we `derive(Default)` on the `InnerSimulationPlugin` and pass the type to our `CrystalOrbClientPlugin` as a generic parameter rather than by value. Then, we can do this in our crystalorb-bevy plugin:

```rust
pub struct CrystalOrbWorld<InnerSimulationPlugin: Plugin + Default, ...> {
    world: BevyWorld,
    schedule: Schedule,
    // ...
}
impl<InnerSimulationPlugin: Plugin + Default> Default for CrystalOrbWorld<InnerSimulationPlugin> {
    fn default() -> Self {
        let app = App::build()
            .add_plugin(InnerSimulationPlugin::default())
            .app;
        Self {
            world: app.world,
            schedule: app.schedule,
            // ...
        }
    }
}
```

This way, I think we could get rid of having the slightly annoying `app_factory` closure. I think for the game developer's perspective, their game code can become something like this:

```rust
#[derive(Default)]
pub struct InnerSimulationPlugin; // Most plugins are an empty struct anyway.

impl Plugin for InnerSimulationPlugin {
    fn build(&mut self, app_builder: AppBuilder) {
        app_builder.add_system(...); // etc...
    }
}

// later on

App::build()
    .add_plugin(OuterPlugin)
    .add_plugin(CrystalOrbClientPlugin::<InnerSimulationPlugin>::new())
    .run();
```

This does add a bit of restriction on what `InnerSimulationPlugin` can be, but I'm guessing for most games that won't be a problem.

Sorry I ended up dumping more ideas to you 😅 . Feel free to ignore it if you've already got a plan or what I said doesn't make sense, but feel free to check this idea out if you're stuck and want some inspiration.

**Comment by Shatur** [comment]
@jamescarterbell are you planning to continue working on it?

**Comment by Shatur** [comment]
@ErnWong are you planning to update your awesome crate to Bevy 0.6?

**Comment by ErnWong** [comment]
@Shatur Sure, I can have a look, maybe this weekend. I've created #26 to track this.
